File: doubleLinkedList/double_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None

    def __str__(self):
        return f'{self.value}'

class DoubleLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def append(self, value):
        node  = Node(value)
        if not self.head:
            self.head = node
            self.tail = node
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node
        self.length += 1
        return self

    def pop(self):
        if not self.head:
            return None

        current = self.tail
        if self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            self.tail = current.prev
            self.tail.next = None
        self.length -= 1
        current.prev = None
        return current

    def shift(self):
        if not self.head:
            return None

        current = self.head
        if self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            self.head = current.next
            self.head.prev = None
        self.length -= 1
        current.next = None
        return current

    def unshift(self, value):
        node = Node(value)
        if not self.head:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head.prev = node
            self.head = node
        self.length += 1
        return self

    def get_node(self, index):
        if not 0 <= index < self.length:
            return None
        if (self.length / 2 > index):
            count = 0
            current = self.head
            while count != index:
                current = current.next
                count += 1
        else:
            count = self.length - 1
            current = self.tail
            while count != index:
                current = current.prev
                count -= 1

        return current

    def set_node(self, index, value):
        current = self.get_node(index)
        if current:
            current.value = value
            return True
        return False

    def insert(self, index, value):
        pass

    def remove(self, value):
        pass

    def __str__(self):
        current = self.head
        arr = []
        while current:
            arr.append(current)
            current = current.next

        return ' <-> '.join(arr)

File: doubleLinkedList/doubleLinkedList.js
class Node {
  constructor(value) {
    this.value = value;
    this.prev = null;
    this.next = null;
  }
}

class DoubleLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  push(value) {
    let node = new Node(value);
    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      node.prev = this.tail;
      this.tail.next = node;
      this.tail = node;
    }
    this.length++;
    return this;
  }

  pop() {
    if (!this.head) return undefined;
    let current = this.tail;
    if (this.head == this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = current.prev;
      this.tail.next = null;
    }
    this.length--;
    current.prev = null;
    return current;
  }

  shift() {
    if (!this.head) return undefined;
    let current = this.head;
    if (this.head == this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = current.next;
      this.head.prev = null;
    }
    this.length--;
    current.next = null;
    return current;
  }

  unshift(value) {
    let node = new Node(value);
    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head.prev = node;
      this.head = node;
    }
    this.length++;
    return this;
  }

  get(index) {
    if (index < 0 || index >= this.length) return null;
    let count;
    let current;
    if (Math.floor(this.length / 2) < index) {
      count = this.length - 1;
      current = this.tail;

      while (count !== index) {
        current = current.prev;
        count--;
      }
    } else {
      count = 0;
      current = this.head;

      while (count !== index) {
        current = current.next;
        count++;
      }
    }
    return current;
  }

  set(index, value) {
    let current = this.get(index);
    if (current) {
      current.value = value;
      return true;
    }
    return false;
  }

  insert(index, value) {
    if (index === 0) {
      return !!this.unshift(value);
    }

    if (index === this.length) {
      return !!this.push(value);
    }

    let current = this.get(index - 1, value);
    if (current) {
      let node = new Node(value);
      node.next = current.next;
      node.next.previus = node;
      node.previus = current;
      current.next = node;
      this.length++;
      return true;
    }
    return false;
  }

  remove(index) {
    if (index === 0) return this.shift();
    if (index === this.length - 1) return this.pop();
    let previousNode = this.get(index - 1);
    if (!previousNode) return undefined;

    let current = previousNode.next;
    previousNode.next = current.next;
    previousNode.next.prev = previousNode;
    this.length--;
    return current;
  }

  reverse() {
    let current = this.head;
    this.head = this.tail;
    this.tail = current;

    while (current) {
      let prev = current.prev;
      let next = current.next;
      current.prev = next;
      current.next = prev;
      current = current.prev;
    }
    return this;
  }
}
